docs/gitbook/tutorials/rollout-weights.md

```diff
@@ -0,0 +1,40 @@
+# Linkerd Rollout Weights
+
+This guide shows you how to use weights in Flagger to have more fine-grained rollouts.  
+
+By default Flagger allows to use linear promotion metrics, with the start value, the step and maximum weight value in 0 to 100 range.  
+
+Example:
+```yaml
+canary:
+  analysis:
+    promotion:
+      maxWeight: 50
+      stepWeight: 20
+```
+This configuration performs analysis starting from 20, increasing by 20 until weight goes above 50.  
+We would have steps (canary weight : primary weight):
+* 20 (20 : 80)
+* 40 (40 : 60)
+* 60 (60 : 40)
+* promotion
+
+In order to enable non-linear promotion a new parameters were introduced:
+* `fullWeight` - determines the sum of canary and primary weights, i.e. the maximum possible weight which can be set to the split
+* `stepWeights` - determines the ordered array of weights, which shall be used during canary promotion.
+
+Example:
+```yaml
+canary:
+  analysis:
+    promotion:
+      fullWeight: 1000
```

Comment:
It's a bit confusing to use different units for custom weights. Did you consider using a normal 0-100 range where the promotion happens at the last defined value?

Reply:
This works with Linkerd for now, but when Linkerd will switch to TrafficSplit v1alpha3 then it needs 0-100. @robq99 wouldn't this break all the other service meshes and ingress controllers? none accepts a wight bigger than 100.

Reply:
> It's a bit confusing to use different units for custom weights. Did you consider using a normal 0-100 range where the promotion happens at the last defined value?
I cannot use 0-100 range as I want to have more granular deployment, like 0.01% => 0.1% => 1%, etc. I would need to change weight to float, which would be more dramatic change (FYI: this is my first code in golang...), if possible at all.

Reply:
That could be a future improvement. Otherwise this creates feature drift where custom weight vs non custom weight strategies use different formats to represent the same thing.

Reply:
> This works with Linkerd for now, but when Linkerd will switch to TrafficSplit v1alpha3 then it needs 0-100. @robq99 wouldn't this break all the other service meshes and ingress controllers? none accepts a weight bigger than 100.

SMI docs states in [smi-example-implementation](https://github.com/servicemeshinterface/smi-spec/blob/master/apis/traffic-split/traffic-split-WD.md#example-implementation) that weight can go above 100.
Of course, if SMI standard does not allow it then the whole idea is broken but as I was not able to find any constraints on weights on both, SMI and Linkerd2 side.
@stefanprodan I saw you are involved into `v1alpha2` migration on Linkerd2 side, could it be updated in a way that any weights are supported? I.e. no 0-100 constraint?

Reply:
> That could be a future improvement. Otherwise this creates feature drift where custom weight vs non custom weight strategies use different formats to represent the same thing.

Do you mean to split implementation of custom weights (like 1,5, 99) and having weights at any arbitrary range (like above 100)?

Reply:
Yes ^

Given that Flagger is a tool that is used primarily to reduce risk I think it would make sense to focus more on the ergonomics and to make it as less error prone as possible.

One option could be using decimals (for example 0.1% could simply be represented as 0.1). I understand that using floats here can be challenging but I think it's worth having a discussion about this topic.

Reply:
I just checked with split.smi-spec.io/v1alpha1/TrafficSplit and:
```yaml
apiVersion:  split.smi-spec.io/v1alpha1
kind:         TrafficSplit
spec:
...
  backends:
    service:  some-service1
    weight:   0.1
    service:  some-service2
    weight:   0.01
```
worked.
@stefanprodan could you please confirm that this will work in v1alpha2/v1alpha3? If yes then I can change the range to 0..100 and convert ints to floats.

Reply:
@robq99 I forgot to update the SMI docs on this, the weight is of type `int` and it should only accept `0-100` range like any other service mesh and ingress out there. Here is the actual code https://github.com/servicemeshinterface/smi-sdk-go/blob/master/pkg/apis/split/v1alpha3/traffic_split.go#L50

Reply:
Is there any particular reason that this is int, not float? Can we potentially change it to float?
As with those two conditions (int, 0-100 range) it is not possible to have more granular rollouts, like 0.01%, 0.1%, etc.

Reply:
@robq99 Envoy doesn't support floats nor does NGINX or any other proxy that I know of except for Linkerd.

Reply:
This is about float vs int. So, float cannot be used, checked.
Can we then use >100 total value? Envoy and NGINX support total weight > 100, not sure about other proxies, though.

Reply:
@stefanprodan I removed `fullWeight` from schemas. I extracted `100` in scheduler to `totalWeight` method - this is not changing any weight related functionality but will make my life easier as I will play with total weight on my fork, I hope this is OK.